pkg/controller/trainingjob_controller.go

```diff
@@ -0,0 +1,258 @@
+	PaddleCli paddleclient.Interface
+
+	trainingjobLister paddlelisters.TrainingJobLister
+	trainingjobSynced cache.InformerSynced
+
+	// jobupdater is
+	jobupdater *sync.Map
+
+	// workqueue is a rate limited work queue. This is used to queue work to be
+	// processed instead of performing it as soon as a change happens.
+	workqueue workqueue.RateLimitingInterface
+	// recorder is an event recorder for recording Event resources to the
+	// Kubernetes API.
+	recorder record.EventRecorder
+}
+
+// New returns a trainingjob controller
+func New(
+	kubeCli kubernetes.Interface,
+	extCli extclient.Interface,
+	paddleCli paddleclient.Interface,
+	paddleInformers paddleinformers.SharedInformerFactory) *TrainingJobController {
+	trainingjobInformer := paddleInformers.Paddlepaddle().V1().TrainingJobs()
+	paddlescheme.AddToScheme(scheme.Scheme)
+
+	glog.V(4).Info("Creating trainingjob event broadcaster")
+	eventBroadcaster := record.NewBroadcaster()
+	eventBroadcaster.StartLogging(glog.Infof)
+	eventBroadcaster.StartRecordingToSink(&typedcorev1.EventSinkImpl{Interface: kubeCli.CoreV1().Events("")})
+	workqueue := workqueue.NewNamedRateLimitingQueue(workqueue.DefaultControllerRateLimiter(), "TrainingJob")
+	recorder := eventBroadcaster.NewRecorder(scheme.Scheme, corev1.EventSource{Component: "TrainingJobController"})
+
+	controller := &TrainingJobController{
+		KubeCli:           kubeCli,
+		ExtCli:            extCli,
+		PaddleCli:         paddleCli,
+		trainingjobLister: trainingjobInformer.Lister(),
+		trainingjobSynced: trainingjobInformer.Informer().HasSynced,
+		jobupdater:        new(sync.Map),
+		workqueue:         workqueue,
+		recorder:          recorder,
+	}
+
+	glog.Info("Setting up event handlers")
+	trainingjobInformer.Informer().AddEventHandler(cache.ResourceEventHandlerFuncs{
+		AddFunc: controller.enqueue,
+		UpdateFunc: func(oldObj, newObj interface{}) {
+			oldTj := oldObj.(*paddlev1.TrainingJob)
+			newTj := newObj.(*paddlev1.TrainingJob)
+			if oldTj.ResourceVersion == newTj.ResourceVersion {
+				glog.V(4).Infof("same resourceversion for training job %s/%s, skipped", oldTj.Namespace, oldTj.Name)
+				return
+			}
+			glog.V(4).Infof("resourceversion for training job %s/%s updated", oldTj.Namespace, oldTj.Name)
+			controller.enqueue(newObj)
+		},
+		DeleteFunc: controller.dequeue,
+	})
+
+	return controller
+}
+
+// Run will set up the event handlers for trainingjob, as well as syncing
+// informer caches and starting workers. It will block until stopCh
+// is closed, at which point it will shutdown the workqueue and wait for
+// workers to finish processing their current work items.
+func (c *TrainingJobController) Run(threadiness int, stopCh <-chan struct{}) error {
+	// TODO add a lock to ensure there is only one controller in the cluster
+	defer runtime.HandleCrash()
+	defer c.workqueue.ShutDown()
+
+	glog.Info("Starting trainingjob controller")
+	glog.Info("Starting to create custom resource definition")
```

Comment:
Maybe can reduce some info logs, output info log only when the operation success or fail.

Reply:
Maybe we can use different log level for them, abundant logs are useful when debugging.